Add toil roles

cket opened the pull request:
Adds ToilBox, Worker, and Leader roles as well as a create-toil-cluster command.

resolves #41, resolves BD2KGenomics/toil#244, resolves BD2KGenomics/toil#245

Commit:
Initial commit of jobtree roles and create-job-tree-cluster command. Functional, but no tests yet

Commit:
Merge branch 'master' into jobtree

Commit:
Run mesos as jobtreebox user instead of mesosbox. This required changes in mesosbox to avoid name mangling and allow jobtreebox to override the methods that interact with the user.

Commit:
EBS with mesos.

Commit:
Specify Mesos version, small changes to jobTree roles to work with EBS

Commit:
Pip install from toil instead of jobtree

hannes-ucsc committed:
Added jobtree subproject to Makefile; added .gitignore to jobtree subproject

Commit:
Remove duplicate code and refactor jobtree to toil.

Commit:
Merge branch 'jobtree' of https://github.com/BD2KGenomics/cgcloud into jobtree

Commit:
Allow user to pass path to a directory to the create-toil command, which will be securely placed on every node in the  cluster

Commit:
Use _list_packages_to_install instead of individual calls to apt-get

hannes-ucsc commented on mesos/src/cgcloud/mesos/mesos_box.py:
```diff
@@ -96,19 +107,36 @@ def _setup_package_repos( self ):
+        role_name += '--' + abreviated_snake_case_class_name( MesosBox )
+        policies.update( dict(
+            ec2_read_only=ec2_read_only_policy,
+            ec2_spark_box=dict( Version="2012-10-17", Statement=[
```
s/spark/mesos/

hannes-ucsc commented on mesos/src/cgcloud/mesos/mesos_box.py:
```diff
@@ -96,19 +107,36 @@ def _setup_package_repos( self ):
 
+    def _list_packages_to_install( self ):
+        return super( MesosBox, self )._list_packages_to_install( ) + [
+            'mesos=0.21.1-1.1.ubuntu1404', 'docker.io']
```
Shouldn't the docker functionality be in ToilBox?

hannes-ucsc replied:
Speaking of which. Maybe we can pull the docker-related stuff into a mixin that John could also use for his Jenkins slave role.

hannes-ucsc commented on mesos/src/cgcloud/mesos/mesos_box.py:
```diff
@@ -214,6 +245,11 @@ def __register_upstart_jobs( self, service_map ):
+    @fabric_task
+    def _docker_group(self):
+        sudo("gpasswd -a {} docker".format(user))
+        sudo("sudo service docker.io restart")
```
Double sudo

hannes-ucsc commented on toil/setup.py:
```diff
@@ -0,0 +1,22 @@
+        'cgcloud-core>=' + cgcloud_version,
+        'bd2k-python-lib>=1.6.dev1',
+        'Fabric>=1.7.0',
+        'lxml>=3.2.1' ] )
```
Should be a newline at the end of the file.

hannes-ucsc commented on toil/src/cgcloud/toil/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from cgcloud.toil.toil_cluster import CreateToilCluster
+
+BOXES = [ ToilBox, ToilLeader, ToilWorker ]
+COMMANDS = [ CreateToilCluster ]
```
Newline at end of file.

hannes-ucsc commented on toil/src/cgcloud/toil/toil_box.py:
```diff
@@ -0,0 +1,108 @@
+
+    def _post_install_packages( self ):
+        super( ToilBox, self )._post_install_packages( )
+        self._install_boto( )
```
This will soon be obsolete when https://github.com/BD2KGenomics/toil/issues/207 is resolved. Maybe add a FIXME?

hannes-ucsc commented on toil/src/cgcloud/toil/toil_box.py:
```diff
@@ -0,0 +1,108 @@
+    def _list_packages_to_install( self ):
+        # packages to apt-get
+        return super( MesosBox, self )._list_packages_to_install( ) + [
+            'git', 'python-dev']
```
git should soon be unnecessary, when https://github.com/BD2KGenomics/toil/issues/215 is resolved. Another FIXME ?

hannes-ucsc commented on toil/src/cgcloud/toil/toil_cluster.py:
```diff
@@ -0,0 +1,33 @@
+from subprocess import check_call
+
+log = logging.getLogger( __name__ )
+shared_dir="/home/ubuntu/shared/"
```
The shared dir should live in ~mesosbox, not ~ubuntu. On the leader and the workers.

hannes-ucsc commented on toil/src/cgcloud/toil/toil_cluster.py:
```diff
@@ -0,0 +1,33 @@
+        dir = options.dir_to_copy
+        if dir:
+            check_call( ['rsync','-r','-e', 'ssh -o StrictHostKeyChecking=no', dir, "ubuntu@"+str(master.ip_address)+":"+shared_dir] )
+        log.info( "=== Launching workers ===" )
```
Can we use the RsyncCommand or Box.rsync() here instead? 

Also, I'd rather have this prompt the user and remove the insecure StrictHostKeyChecking=no option. I plan on fixing host key checking on a global basis and leaving that option here would mean that this code wouldn't benefit from my fix in the future.

hannes-ucsc commented on toil/src/cgcloud/toil/toil_cluster.py:
```diff
@@ -0,0 +1,33 @@
+        log.info( "=== Launching leader ===" )
+        if options.instance_type is None:
+            options.instance_type = options.slave_instance_type
+        return self.run_on_box( options, ToilLeader( ctx) )
```
Add newline

hannes-ucsc commented on toil/src/cgcloud/toil/toil_cluster.py:
```diff
@@ -0,0 +1,33 @@
+class CreateToilCluster(CreateMesosCluster):
+    def __init__( self, application ):
+        super( CreateToilCluster, self ).__init__( application )
+        self.option( '--dir_to_copy', default=None,
```
What do you think about using `--shared-dir` instead?

In either case, it should be using dashes not underscores, IIRC.

Commit:
Various minor improvements in mesos and toil roles

Move docker functionality to ToilBox, move shared dir to mesosbox's home dir, document future changes with FIXMEs, use the box.rsync method, add newlines at the end of files.

Commit:
Improper super call caused mesos install to be skipped

Commit:
Install toil from pypi instead of git

Commit:
Remove a now unnecessary chmod, upgrade pip on box.

Commit:
Merge branch 'master' into jobtree

Conflicts:
	mesos/src/cgcloud/mesos/mesos_box.py

Commit:
Have mesos, toil write to /var/lib/mesos and /var/lib/toil respectively

hannes-ucsc commented on mesos-tools/src/cgcloud/mesos_tools/__init__.py:
```diff
@@ -22,14 +22,19 @@
 
-class MesosTools( object ):
+work_dirs = ['/var/lib/mesos', '/var/lib/toil']
 
```
The `mesos` subproject can't reference anything Toil-specific.

I have the feeling both shared_dir and work_dirs should be parameters to the MesosTools constructor.

When defining a constant, there should only be one definitions, not two or more that have to be kept in sync. As a very last resort (not here though) all definitions have to be annotate with a comment that clearly states where the other definitions are and that they need to be kept in sync.

Commit:
Add missing '-r' for --shared-dir rsync call

Commit:
Remove reference to toil role in mesos
